### PR TITLE
[analytics-engine] Coverage fixes for search command on the analytics-engine route

### PR DESCRIPTION
**Pull request by ahkcs**
### What this PR does

Four focused fixes in the analytics-engine sandbox plugins, narrowly scoped to the `search` command's analytics-engine route.

| # | Commit | What it fixes |
|---|---|---|
| 1 | `Register TIMESTAMP in STANDARD_PROJECT_OPS` | PPL `timestamp(expr)` (also implicit when the analyzer coerces a string literal to TIMESTAMP for `@timestamp="..."` comparisons) was rejected by `OpenSearchProjectRule` even though `TimestampFunctionAdapter` already wires it to DataFusion's `to_timestamp`. |
| 2 | `Catch Calcite Litmus.THROW AssertionError in DefaultPlanExecutor` | Calcite's `Litmus.THROW` raises raw `AssertionError` from non-asserted Java. Without a catch, the analytics-engine executor lets it bubble to `OpenSearchUncaughtExceptionHandler` and kills the cluster JVM. Mirrors the SQL-plugin-side catch in `UnifiedQueryPlanner.plan`. |
| 3 | `Preserve fractional seconds in DatetimeOutputCastRewriter` | Widens the `to_char` format from `'%Y-%m-%d %H:%M:%S'` to `'%Y-%m-%d %H:%M:%S%.f'` so `date_nanos` columns keep sub-second precision in the response. Follow-up to [#21650](https://github.com/opensearch-project/OpenSearch/pull/21650). |
| 4 | `Apply schema coercion on indexed-executor placeholder` | The indexed-execution path inferred the parquet schema via `build_segments` and registered it directly on the `PlaceholderProvider` that `from_substrait_plan` binds against — bypassing the `schema_coerce::coerce_inferred_schema` that the three other `infer_schema` sites (session_context, query_executor, api) all apply. So parquet's `BinaryView` for `ip` columns leaked to the substrait bind step, conflicting with isthmus's `Binary`. One-line fix: call `coerce_inferred_schema` on the schema returned by `build_segments`. |

### Why commit 4 matters

Without commit 4, every analytics-engine query against an OTEL-logs-style index (any mapping that includes an `ip` field) fails at fragment start with:

```
Substrait error: Field 'attributes.client.ip' in Substrait schema has a different type
(Binary) than the corresponding field in the table schema (BinaryView).
```

The Substrait plan declares `Binary` (isthmus has no view types), the table provider reports `BinaryView` (parquet's storage form). The three other `infer_schema` call sites coerce; only the indexed-executor path slipped. Adding the missing coerce call restores parity.

### Pass / fail breakdown

`CalciteSearchCommandIT` on the analytics-engine route, against current `upstream/main`:

| Step | PASS / 52 |
|---|---|
| Pre-#21631 baseline | 3 |
| + #21631 (IP/BINARY/MATCH_ONLY_TEXT scan + `schema_coerce` BinaryView↔Utf8) | ~24 |
| + #21698 + #21628 + #21622 + #21562 (unsupported-type drop + PPL conversion + dedup + relevance functions on Lucene secondary) | 27 (indexed-executor schema_coerce gap surfaces here) |
| + Commit 4 (schema_coerce on indexed-executor) | 34 (+7 — every OTEL-logs query that scans an `ip` column unblocked at the bind step) |
| + Commits 1, 2, 3 (TIMESTAMP capability, Litmus catch, fractional seconds) | 34 (no new passes; safety / precision improvements that the existing test queries don't directly assert on) |
| + [opensearch-project/sql#5447](https://github.com/opensearch-project/sql/pull/5447) (correctness fixes: datetime schema-type recovery + IP `byte[]` → `ExprIpValue`) | **26** |

The drop from 34 → 26 with sql#5447 is **not a regression**: sql#5447's earlier iteration carried an SQL-side workaround (lower numeric / boolean predicates to typed `RexCall` to skip `query_string`) that lifted the same suite to 38 / 52. That workaround was deliberately removed in the latest sql#5447 revision because the root cause — **Lucene-secondary `query_string` doesn't re-type literals inside the query body using the field mapping** — should be fixed at the Lucene-secondary layer rather than in every consumer.

### Remaining 25 failures — single upstream root cause, not blocking this PR

All 25 collapse to the typed-literal `query_string` gap on the analytics-engine Lucene-secondary path (adjacent to [opensearch-project/OpenSearch#21562](https://github.com/opensearch-project/OpenSearch/pull/21562)):

- 5 numeric / decimal range tests
- 10 date-math / date range tests
- 4 compound-predicate tests (mixed-AND, same family as #21562's tests 10c / 10d that the author flagged as "handled separately")
- 2 attribute / multi-field tests
- 1 IP-via-`query_string` test
- 2 Lucene-secondary edge cases (wildcard pattern, NOT semantics)
- 1 composite-engine empty-mapping streaming test

See [opensearch-project/sql#5447](https://github.com/opensearch-project/sql/pull/5447) for the per-test bucketing and a minimal smoking-gun repro:

```
where query_string([severityText], "ERROR")               # ✅ works
where query_string([severityNumber], "severityNumber:>15") # ❌ 0 rows (expected 6)
```

### Testing

- **Sandbox-wide check**: `./gradlew check -p sandbox -Dsandbox.enabled=true` — see CI
- **`CalciteSearchCommandIT` on analytics route**: 26 / 52 pass with this PR + sql#5447 against current `upstream/main`

### CODE REVIEW

File: sandbox/plugins/analytics-backend-datafusion/src/main/java/org/opensearch/be/datafusion/DataFusionAnalyticsBackendPlugin.java [hunk 1]
```diff
@@ -62,6 +62,9 @@ public class DataFusionAnalyticsBackendPlugin implements AnalyticsSearchBackendP
         SUPPORTED_FIELD_TYPES.addAll(FieldType.date());
         SUPPORTED_FIELD_TYPES.add(FieldType.BOOLEAN);
         SUPPORTED_FIELD_TYPES.add(FieldType.TEXT);
+        SUPPORTED_FIELD_TYPES.add(FieldType.BINARY);
```
Comment (vinaykpud) [inline sandbox/plugins/analytics-backend-datafusion/src/main/java/org/opensearch/be/datafusion/DataFusionAnalyticsBackendPlugin.java:70]: This is already added as part of https://github.com/opensearch-project/OpenSearch/pull/21631

File: sandbox/libs/analytics-api/src/main/java/org/opensearch/analytics/schema/OpenSearchSchemaBuilder.java [hunk 2]
```diff
@@ -146,6 +150,16 @@ private static void addLeafFields(
...
                 continue;
             }
+            // TEMP: IP columns surface in DataFusion as Arrow BinaryView while the Calcite
+            // VARCHAR mapping below makes Substrait declare Utf8. The schema mismatch fails
+            // every scan over an index containing any ip-typed leaf — even when the query
+            // doesn't project that column. Skipping ip fields here unblocks the rest of the
+            // schema until a proper BinaryView ↔ Utf8 conversion lands in the DataFusion
+            // scan path; queries that explicitly reference the column will surface as "field
+            // not found" instead of the much-deeper Substrait-validation crash.
+            if ("ip".equals(fieldType)) {
```
Comment (vinaykpud) [inline sandbox/libs/analytics-api/src/main/java/org/opensearch/analytics/schema/OpenSearchSchemaBuilder.java]: we are supporting this after https://github.com/opensearch-project/OpenSearch/pull/21631

File: sandbox/plugins/analytics-backend-datafusion/src/main/java/org/opensearch/be/datafusion/DataFusionAnalyticsBackendPlugin.java (hunk 1 above)
Comment (ahkcs) [reply sandbox/plugins/analytics-backend-datafusion/src/main/java/org/opensearch/be/datafusion/DataFusionAnalyticsBackendPlugin.java:70]: Confirmed — #21631 added IP/BINARY/MATCH_ONLY_TEXT to `SUPPORTED_FIELD_TYPES` directly. I dropped this redundant commit during the rebase onto current `main`; thanks for the pointer.

File: sandbox/libs/analytics-api/src/main/java/org/opensearch/analytics/schema/OpenSearchSchemaBuilder.java (hunk 2 above)
Comment (ahkcs) [reply sandbox/libs/analytics-api/src/main/java/org/opensearch/analytics/schema/OpenSearchSchemaBuilder.java]: Right — #21631 maps `ip` → `VARBINARY` properly (with `schema_coerce.rs` handling BinaryView↔Utf8 at the scan boundary), so the temp ip-skip I had here is no longer needed. #21698 also generalised the unknown-type drop, which subsumes the date_nanos workaround commit. Both commits dropped during the rebase.

File: sandbox/plugins/analytics-backend-datafusion/src/main/java/org/opensearch/be/datafusion/DataFusionAnalyticsBackendPlugin.java [hunk 3]
```diff
@@ -228,11 +228,15 @@ public class DataFusionAnalyticsBackendPlugin implements AnalyticsSearchBackendP
         // see a real time/date type and Isthmus serializes accordingly.
         ScalarFunction.TIME,
         ScalarFunction.DATE,
+        // PPL `timestamp(expr)` — parse a string into a TIMESTAMP, also used by
+        // the analyzer to coerce string literals when comparing against TIMESTAMP
+        // columns (e.g. `@timestamp="2024-01-15T..."`). The
+        // TimestampFunctionAdapter already wires the call shape; the missing
+        // capability entry was blocking the project rule from accepting the
+        // operator even though DataFusion executes it via `to_timestamp`.
```
Comment (vinaykpud) [inline sandbox/plugins/analytics-backend-datafusion/src/main/java/org/opensearch/be/datafusion/DataFusionAnalyticsBackendPlugin.java]: Can we make this comment short or remove? This comment mostly seems like explains why we added `ScalarFunction.TIMESTAMP` here, which is implicit., that to support timestamp in projection.

File: sandbox/plugins/analytics-engine/src/main/java/org/opensearch/analytics/exec/DefaultPlanExecutor.java [hunk 4]
```diff
@@ -113,6 +113,14 @@ public void execute(RelNode logicalFragment, Object context, ActionListener<Iter
                 executeInternal(logicalFragment, listener);
             } catch (Exception e) {
                 listener.onFailure(e);
+            } catch (AssertionError e) {
```
Comment (vinaykpud) [inline sandbox/plugins/analytics-engine/src/main/java/org/opensearch/analytics/exec/DefaultPlanExecutor.java:121]: Since the `Exception` is declared before `AssertionError` will this code path will be hit?

Comment (mch2) [reply sandbox/plugins/analytics-engine/src/main/java/org/opensearch/analytics/exec/DefaultPlanExecutor.java:121]: if the assertion comes from an inline assert, this should only hit in tests

File: sandbox/plugins/analytics-backend-datafusion/rust/src/indexed_executor.rs [hunk 5]
```diff
@@ -444,6 +444,17 @@ pub async unsafe fn execute_indexed_with_context(
     let (segments, schema) = build_segments(&state, Arc::clone(&store), object_metas.as_ref(), writer_generations.as_ref())
         .await
         .map_err(DataFusionError::Execution)?;
+    // Substrait's type system is narrower than Arrow's; normalize the inferred schema to
```
Comment (vinaykpud) [inline sandbox/plugins/analytics-backend-datafusion/rust/src/indexed_executor.rs]: do we need this comment?

File: sandbox/plugins/analytics-engine/src/main/java/org/opensearch/analytics/exec/DefaultPlanExecutor.java (hunk 4 above)
Comment (sandeshkr419) [reply sandbox/plugins/analytics-engine/src/main/java/org/opensearch/analytics/exec/DefaultPlanExecutor.java:121]: @mch2 not exactly - it will go to the base class and not hit this at all - what did I miss?

Comment (mch2) [reply sandbox/plugins/analytics-engine/src/main/java/org/opensearch/analytics/exec/DefaultPlanExecutor.java:121]: got it this is coming from litmus - we should be defensive here and catch or it will cause local node crash.

File: sandbox/plugins/analytics-backend-datafusion/rust/src/indexed_executor.rs (hunk 5 above)
Comment (ahkcs) [reply sandbox/plugins/analytics-backend-datafusion/rust/src/indexed_executor.rs]: Done — removed in the latest force-push.

File: sandbox/plugins/analytics-backend-datafusion/src/main/java/org/opensearch/be/datafusion/DataFusionAnalyticsBackendPlugin.java (hunk 3 above)
Comment (ahkcs) [reply sandbox/plugins/analytics-backend-datafusion/src/main/java/org/opensearch/be/datafusion/DataFusionAnalyticsBackendPlugin.java]: Done — removed in the latest force-push.

**Review by mch2 — APPROVED** [review]

**Review by vinaykpud — APPROVED** [review]
